### PR TITLE
Add spark mapstate

### LINKED ISSUE #31668: [Feature Request]: Add MapState in SparkRunner
### What would you like to happen?

Currently, SparkRunner does not support MapState states.
While more types of states will be added in the future, SparkRunner only supports ValuleState, BagState, CombiningState, and WatermarkHoldState.

Therefore, we need to support the types of states that we don't currently support.

### Issue Priority

Priority: 2 (default / most feature requests should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [X] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by twosom**
**Please** add a meaningful description for your change here

fixes #31668 

This PR contains theses change
- add `isEmpty` test in `StateInternalsTest#testMap`
- add map state in `SparkRunner`
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by twosom** [comment]
retest this please

**Comment by twosom** [comment]
retest this please

**Comment by twosom** [comment]
retest this please

**Comment by chamikaramj** [comment]
assign set of reviewers

**Comment by twosom** [comment]
> Thanks for the contribution. Could you please update the comments of the test class here
> 
> 
> 
> https://github.com/apache/beam/blob/54a1538ada2161dc5ae848252d3fb06aaacfc001/runners/spark/src/test/java/org/apache/beam/runners/spark/stateful/SparkStateInternalsTest.java#L27
> 
> 
> 
> now it no longer ignores map tests

I hadn't seen that comment, thanks!

**Review by Abacn — APPROVED** [review]
thanks, just have a minor comment

when pushing changes, please do not force push and just push a new commit instead, so reviewer can track diff

**Comment by twosom** [comment]
> thanks, just have a minor comment
> 
> when pushing changes, please do not force push and just push a new commit instead, so reviewer can track diff

Got it! I'll make sure to push new commits from now on, instead of force-pushing. Thanks so much!

File: runners/core-java/src/test/java/org/apache/beam/runners/core/StateInternalsTest.java [hunk 1]
```diff
@@ -386,10 +386,16 @@ public void testMap() throws Exception {
         value.entries().readLater().read(),
         containsInAnyOrder(MapEntry.of("B", 2), MapEntry.of("D", 4), MapEntry.of("E", 5)));
 
+    // isEmpty
+    assertThat(value.isEmpty().read(), equalTo(false));
```
[the remark this answers is not in the document]
Comment (twosom) [reply runners/core-java/src/test/java/org/apache/beam/runners/core/StateInternalsTest.java:390]: check!

**Pull request merged**